src/lib.rs

```diff
@@ -1128,6 +1269,25 @@ impl Bump {
     }
 }
 
+/// Writes `f`'s return value to `ptr`.
+#[inline(always)]
+unsafe fn alloc_with_writer<T, F>(ptr: *mut T, f: F)
+where
+    F: FnOnce() -> T,
+{
+    // This function is translated as:
+    // - allocate space for a T on the stack
+    // - call f() with the return value being put onto this stack space
+    // - memcpy from the stack to the heap
+    //
+    // Ideally we want LLVM to always realize that doing a stack
+    // allocation is unnecessary and optimize the code so it writes
+    // directly into the heap instead. It seems we get it to realize
+    // this most consistently if we put this critical line into it's
+    // own function instead of inlining it into the surrounding code.
+    ptr::write(ptr, f())
+}
```

Comment:
I moved this function out of of the others to deduplicate it, but I don't think this would make a difference after compiling them. I'd like you to decide where to place them.